integration/library_dockermode_test.go

```diff
@@ -0,0 +1,186 @@
+			expectedOS:    analyzer.OS{Name: "7.7", Family: "redhat"},
+		},
+		{
+			name:          "happy path, opensuse leap 15.1",
+			imageName:     "opensuse/leap:latest",
+			imageFile:     "testdata/fixtures/opensuse-leap-151.tar.gz",
+			expectedFiles: []string{"usr/lib/os-release", "usr/lib/sysimage/rpm/Packages", "/config", "etc/os-release"},
+			expectedOS:    analyzer.OS{Name: "15.1", Family: "opensuse.leap"},
+		},
+		{
+			name:                 "happy path, vulnimage with lock files",
+			imageName:            "knqyf263/vuln-image:1.2.3",
+			imageFile:            "testdata/fixtures/vulnimage.tar.gz",
+			expectedFiles:        []string{"etc/os-release", "node-app/package-lock.json", "python-app/Pipfile.lock", "ruby-app/Gemfile.lock", "rust-app/Cargo.lock", "/config", "etc/alpine-release", "lib/apk/db/installed", "php-app/composer.lock"},
+			expectedOS:           analyzer.OS{Name: "3.7.1", Family: "alpine"},
+			expectedLibraries:    "testdata/goldens/knqyf263vuln-image:1.2.3.expectedlibs.golden",
+			expectedPkgsFromCmds: "testdata/goldens/knqyf263vuln-image:1.2.3.expectedpkgsfromcmds.golden",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc // save a copy of tc for use in t.Run https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			d, _ := ioutil.TempDir("", "TestFanal_Library_*")
+			defer os.RemoveAll(d)
+			c := cache.Initialize(d)
+
+			opt := types.DockerOption{
+				Timeout:  600 * time.Second,
+				SkipPing: true,
+			}
+
+			cli, err := client.NewClientWithOpts(client.FromEnv)
+			require.NoError(t, err, tc.name)
+
+			// ensure image doesnt already exists
+			_, _ = cli.ImageRemove(ctx, tc.imageFile, dtypes.ImageRemoveOptions{
+				Force:         true,
+				PruneChildren: true,
+			})
+
+			testfile, err := os.Open(tc.imageFile)
+			require.NoError(t, err)
+
+			// load image into docker engine
+			_, err = cli.ImageLoad(ctx, testfile, true)
+			require.NoError(t, err, tc.name)
+
+			// tag our image to something unique
+			err = cli.ImageTag(ctx, tc.imageName, tc.imageFile)
+			require.NoError(t, err, tc.name)
+
+			ext, err := docker.NewDockerExtractor(opt, c)
+			require.NoError(t, err, tc.name)
+			ac := analyzer.Config{Extractor: ext}
+
+			// run tests twice, one without cache and with cache
+			for i := 1; i <= 2; i++ {
+				runChecksDockerMode(t, ac, ctx, tc, d, c)
+			}
+
+			// clear Cache
+			require.NoError(t, c.Clear(), tc.name)
+
+			// remove Image
+			_, err = cli.ImageRemove(ctx, tc.imageFile, dtypes.ImageRemoveOptions{
+				Force:         true,
+				PruneChildren: true,
+			})
+			assert.NoError(t, err, tc.name)
+		})
+	}
+}
+
+func runChecksDockerMode(t *testing.T, ac analyzer.Config, ctx context.Context, tc testCase, d string, c cache.Cache) {
+	actualFiles, err := ac.Analyze(ctx, tc.imageFile)
```

Comment:
discussion: The difference between tar mode and docker mode is almost only this line. What do you think about aggregating those tests?


```
for _, fromFile := []bool{false, true} {
    var actualFiles extractor.FileMap
    var err error

    if fromFile {
        testfile, err := os.Open(tc.imageFile)
        require.NoError(t, err)
        actualFiles, err = ac.AnalyzeFile(ctx, testfile)
    } else {
        actualFiles, err = ac.Analyze(ctx, tc.imageFile)
    }
}
```

I know this is not a good example because we have dockerless mode as well, but anyway I think we can remove duplicate code, especially testcases. I'm not sure if the aggregation is a better way, but I'm afraid to forget to add a new test case in either mode.

Reply:
Yeah I agree. I have addressed this concern in the next commits. I think I'm happy with having a single file and a common set of use cases as seen here: https://github.com/aquasecurity/fanal/pull/76/commits/41881b073d3f0bb9ba2e57069a7e3f7fbaebcf0f

Reply:
Also this way we can run docker mode and tar mode in parallel.